### PR TITLE
Add ability to use Q5_0, Q5_1, and IQ4_NL for quantized K cache

### PR DESCRIPTION
**Pull request by ikawrakow**
Tested to work on CUDA and Metal (CPU worked before this PR).

Table shows PPL for LLaMA-v2-7B quantized with `Q4_K_S` using the available quantization types for the `K` cache. The "K cache size" column is for a context of  4096 tokens:

| K cache type | K cache size | PPL |
|--:|--:|--:|
| fp16 | 1024 MB | 5.8671 |
|Q8_0 | 544 MB | 5.8681 |
|Q5_1 | 384 MB | 5.8802 |
|Q5_0 | 352 MB | 5.8920 |
| Q4_1 | 320 MB | 5.9233 |
|IQ4_NL | 288 MB | 5.9323 |
|Q4_0 | 288 MB | 5.9790 |

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]

File: common/common.cpp [hunk 1]
```diff
@@ -1590,6 +1590,9 @@ static ggml_type kv_cache_type_from_str(const std::string & s) {
     if (s == "q4_1") {
         return GGML_TYPE_Q4_1;
     }
+    if (s == "iq4_nl") {
+        return GGML_TYPE_IQ4_NL;
+    }
```
Comment (slaren) [inline common/common.cpp:1595]: This would also need to be added to `llama-bench` for it to support this type.

**Review by slaren — APPROVED** [review]

File: ggml-metal.m [hunk 2]
```diff
@@ -739,6 +741,9 @@ static bool ggml_metal_supports_op(const struct ggml_metal_context * ctx, const
                            case GGML_TYPE_Q8_0:
                            case GGML_TYPE_Q4_0:
                            case GGML_TYPE_Q4_1:
+                           case GGML_TYPE_Q5_0:
+                           case GGML_TYPE_Q5_1:
+                           case GGML_TYPE_IQ4_NL:
```
Comment (slaren) [inline ggml-metal.m:746]: This also needs to be added to the `supports_op` of the CUDA backend.

**Comment by sorasoras** [comment]
What is the performance of these quants

**Comment by ikawrakow** [comment]
> What is the performance of these quants

For LLaMA-v2-7B quantized with `Q4_K_S` I get

* CUDA (running on RTX-4080)

| model                          | backend    | ngl |    threads | type_k     | test       |              t/s |
| ------------------------------ | ---------- | --: | ---------: | ---------- | ---------- | ---------------: |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | f16        | tg 128     |    127.57 ± 0.59 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | q8_0       | tg 128     |     76.93 ± 0.26 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | q5_1       | tg 128     |     76.66 ± 0.00 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | q5_0       | tg 128     |     76.77 ± 0.02 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | q4_1       | tg 128     |     77.35 ± 0.01 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | iq4_nl     | tg 128     |     75.07 ± 0.36 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | q4_0       | tg 128     |     76.86 ± 0.00 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | f16        | pp 512     |  5675.11 ± 10.20 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | q8_0       | pp 512     |   5209.73 ± 4.39 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | q5_1       | pp 512     |   5203.32 ± 6.13 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | q5_0       | pp 512     |   5200.40 ± 4.07 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | q4_1       | pp 512     |   5197.36 ± 3.78 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | iq4_nl     | pp 512     |   4727.66 ± 0.86 |
| llama 7B Q4_K - Small          | CUDA       | 100 |          1 | q4_0       | pp 512     |   5199.24 ± 1.48 |


* Metal (running on 30-core M2 Max GPU)

| model                          | backend    | ngl |    threads | type_k     | test       |              t/s |
| ------------------------------ | ---------- | --: | ---------: | ---------- | ---------- | ---------------: |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | f16        | tg 128     |     54.13 ± 0.19 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | q8_0       | tg 128     |     53.05 ± 0.20 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | q5_1       | tg 128     |     52.96 ± 0.10 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | q5_0       | tg 128     |     53.24 ± 0.05 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | q4_1       | tg 128     |     53.45 ± 0.07 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | iq4_nl     | tg 128     |     51.53 ± 0.08 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | q4_0       | tg 128     |     53.41 ± 0.11 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | f16        | pp 512     |    474.85 ± 3.24 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | q8_0       | pp 512     |    474.59 ± 3.27 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | q5_1       | pp 512     |    473.34 ± 3.52 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | q5_0       | pp 512     |    472.48 ± 3.57 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | q4_1       | pp 512     |    475.29 ± 2.37 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | iq4_nl     | pp 512     |    468.29 ± 3.38 |
| llama 7B Q4_K - Small          | Metal      | 100 |          4 | q4_0       | pp 512     |    475.45 ± 2.10 |


The `IQ4_NL` quantization used for quantizing the `K` cache is simpler than `IQ4_NL` used for weight quantization to keep the performance drop acceptable.

So, in short, performance is OK-ish on Metal, where we need K-cache quantization less, but quite horrible on CUDA, where we need it more.

**Comment by ggerganov** [comment]
> The IQ4_NL quantization used for quantizing the K cache is simpler than IQ4_NL used for weight quantization to keep the performance drop acceptable.

This makes the `test-backend-ops` fail because the results of the CPY operation betweeb the CPU and GPU differ:

```
  CPY(type_src=f32,type_dst=iq4_nl,ne=[256,4,4,4]): [CPY] NMSE = 0.004881965 > 0.000000100 FAIL
```

@ikawrakow How much is the performance drop with the original implementation?

**Comment by ikawrakow** [comment]
No, simply change `quantize_row_iq4_nl()`, which is not used during quantization but is the `from_float` type trait for `IQ4_NL`, to do the same thing as CUDA/Metal.

**Comment by ggerganov** [comment]
Yes, good idea

**Comment by Artefact2** [comment]
Q{4,5}_{0,1} seem broken on AMD. `perplexity` returns nan.

```
./perplexity --chunks 1 -f wiki.test.raw -c 4096 -ngl 999 -ctk q4_0 -m LLaMA2-13B-Q4_0.gguf
perplexity: tokenizing the input ..
perplexity: tokenization took 482.161 ms
perplexity: calculating perplexity over 1 chunks, n_ctx=4096, batch_size=2048, n_seq=1
perplexity: 11.25 seconds per pass - ETA 0.18 minutes
[1]nan,
Unexpected negative standard deviation of log(prob)
```

**Comment by ikawrakow** [comment]
> Q{4,5}_{0,1} seem broken on AMD. perplexity returns nan.

Did I break it, or was it broken before I added `Q5_0/1` and `IQ4_NL` ?

**Comment by Artefact2** [comment]
It was already broken before.
